libs/api/ebpf_api.cpp

```diff
@@ -4749,6 +4752,205 @@ ebpf_ring_buffer_map_unsubscribe(_In_ _Post_invalid_ ring_buffer_subscription_t*
+          sample_callback(nullptr), lost_callback(nullptr), buffer(nullptr), cpu_id(0), reply({}),
+          async_ioctl_completion(nullptr), async_ioctl_failed(false)
+    {
+    }
+    ~_ebpf_perf_event_array_subscription() { EBPF_LOG_ENTRY(); }
+    std::mutex lock;
+    _Write_guarded_by_(lock) boolean unsubscribed;
+    ebpf_handle_t perf_event_array_map_handle;
+    void* callback_context;
+    perf_buffer_sample_fn sample_callback;
+    perf_buffer_lost_fn lost_callback;
+    uint8_t* buffer;
+    uint32_t cpu_id;
+    ebpf_operation_map_async_query_reply_t reply;
+    _Write_guarded_by_(lock) async_ioctl_completion_t* async_ioctl_completion;
+    _Write_guarded_by_(lock) bool async_ioctl_failed;
+} ebpf_perf_event_array_subscription_t;
+
+typedef std::unique_ptr<ebpf_perf_event_array_subscription_t> ebpf_perf_event_array_subscription_ptr;
+
+static ebpf_result_t
+_ebpf_perf_event_array_map_async_query_completion(_Inout_ void* completion_context) NO_EXCEPT_TRY
+{
+    EBPF_LOG_ENTRY();
+    ebpf_assert(completion_context);
+
+    ebpf_perf_event_array_subscription_t* subscription =
+        reinterpret_cast<ebpf_perf_event_array_subscription_t*>(completion_context);
+
+    const uint32_t cpu_id = subscription->cpu_id;
+
+    size_t consumer = 0;
+    size_t producer = 0;
+    size_t lost_count = 0;
+
+    ebpf_result_t result = EBPF_SUCCESS;
+    // Check the result of the completed async IOCTL call.
+    result = get_async_ioctl_result(subscription->async_ioctl_completion);
+
+    if (result != EBPF_SUCCESS) {
+        if (result != EBPF_CANCELED) {
+            // The async IOCTL was not canceled, but completed with a failure status. Mark the subscription object as
+            // such, so that it gets freed when the user eventually unsubscribes.
+            std::scoped_lock lock{subscription->lock};
+            subscription->async_ioctl_failed = true;
+            EBPF_RETURN_RESULT(result);
+        } else {
+            // User has canceled subscription. Invoke user specified callback for the final time with NULL record. This
+            // will let the user app clean up its state.
+            TraceLoggingWrite(
+                ebpf_tracelog_provider,
+                EBPF_TRACELOG_EVENT_GENERIC_MESSAGE,
+                TraceLoggingLevel(WINEVENT_LEVEL_INFO),
+                TraceLoggingKeyword(EBPF_TRACELOG_KEYWORD_API),
+                TraceLoggingString(
+                    __FUNCTION__, "perf_event_array map async query completion invoked with EBPF_CANCELED."));
+
+            subscription->sample_callback(subscription->callback_context, cpu_id, nullptr, 0);
+        }
+    } else {
+        // Async IOCTL operation returned with success status. Read the ring buffer records and indicate it to the
+        // subscriber.
+
+        size_t perf_event_array_size = 0;
+        uint32_t dummy;
+
+        result = _get_map_descriptor_properties(
```

Comment:
is it possible to cache the map descriptor in `ebpf_perf_event_array_subscription_t` avoid a possible IOCTL call to get map properties each time?